### PR TITLE
Work-around bug in recent Ipopt_jll releases

### PR DESCRIPTION
**Pull request by odow**
This is mainly a test to see if there are any CI problems with AmplNLWRiter.

cc @Vaibhavdixit02

### CODE REVIEW

**Comment by odow** [comment]
cc @amontoison Ipopt_jll is failing?

**Comment by odow** [comment]
Here's the case from @Vaibhavdixit02

```Julia
julia> Ipopt_jll.amplexe() do exe
       run(`$exe test.nl`)
       end
/Users/oscar/.julia/artifacts/6a0533fbc1b71b4c136e88c8407282c6eda28446/bin/ipopt: jacdim: got M = 0, N = 0, NO = 0
ERROR: failed process: Process(`/Users/oscar/.julia/artifacts/6a0533fbc1b71b4c136e88c8407282c6eda28446/bin/ipopt test.nl`, ProcessExited(1)) [1]

Stacktrace:
  [1] pipeline_error
    @ ./process.jl:565 [inlined]
  [2] run(::Cmd; wait::Bool)
    @ Base ./process.jl:480
  [3] run
    @ ./process.jl:477 [inlined]
  [4] (::var"#3#4")(exe::String)
    @ Main ./REPL[5]:2
  [5] (::JLLWrappers.var"#2#3"{var"#3#4", String})()
    @ JLLWrappers ~/.julia/packages/JLLWrappers/QpMQW/src/runtime.jl:49
  [6] withenv(::JLLWrappers.var"#2#3"{var"#3#4", String}, ::Pair{String, String}, ::Vararg{Pair{String, String}})
    @ Base ./env.jl:172
  [7] withenv_executable_wrapper(f::Function, executable_path::String, PATH::String, LIBPATH::String, adjust_PATH::Bool, adjust_LIBPATH::Bool)
    @ JLLWrappers ~/.julia/packages/JLLWrappers/QpMQW/src/runtime.jl:48
  [8] invokelatest(::Any, ::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Base ./essentials.jl:729
  [9] invokelatest(::Any, ::Any, ::Vararg{Any})
    @ Base ./essentials.jl:726
 [10] amplexe(f::Function; adjust_PATH::Bool, adjust_LIBPATH::Bool)
    @ Ipopt_jll ~/.julia/packages/JLLWrappers/QpMQW/src/products/executable_generators.jl:21
 [11] amplexe(f::Function)
    @ Ipopt_jll ~/.julia/packages/JLLWrappers/QpMQW/src/products/executable_generators.jl:19
 [12] top-level scope
    @ REPL[5]:1

shell> cat test.nl

g3 1 1 0
 4 2 1 2 0 0
 2 1
 0 0
 4 4 4
 0 0 0 1
 0 0 0 0 0
 8 4
 0 0
 0 0 0 0 0
C0
o2
o2
o2
v0
v1
v2
v3
C1
o0
o0
o0
o5
v0
n2
o5
v1
n2
o5
v2
n2
o5
v3
n2
O0 0
o0
v2
o2
o2
v0
v3
o0
o0
v0
v1
v2
x4
0 1
1 5
2 5
3 1
r
0 25 Inf
0 40 40
b
0 1 5
0 1 5
0 1 5
0 1 5
k3
2
4
6
J0 4
0 0
1 0
2 0
3 0
J1 4
0 0
1 0
2 0
3 0
G0 4
0 0
1 0
2 0
3 0

(ip) pkg> st
Status `/private/tmp/ip/Project.toml`
  [9cc047cb] Ipopt_jll v300.1400.1000+0
```

I guess this is a recent change in Ipopt_jll.

**Comment by odow** [comment]
Actually @amontoison can you take a look at this? I think it is the new LBT stuff:
```
shell> cat test.nl
g3 1 1 0
 1 0 1 0 0 0
 0 1
 0 0
 0 1 0
 0 0 0 1
 0 0 0 0 0
 0 1
 0 0
 0 0 0 0 0
O0 0
o2
v0
v0
x1
0 0
b
3
G0 1
0 0

julia> import Ipopt_jll

julia> Ipopt_jll.amplexe() do exe
           run(`$exe test.nl`)
       end


******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.4, running with linear solver MUMPS 5.4.1.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        1

Total number of variables............................:        1
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  0.0000000e+00 0.00e+00 0.00e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0

Number of Iterations....: 0

                                   (scaled)                 (unscaled)
Objective...............:   0.0000000000000000e+00    0.0000000000000000e+00
Dual infeasibility......:   0.0000000000000000e+00    0.0000000000000000e+00
Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   0.0000000000000000e+00    0.0000000000000000e+00


Number of objective function evaluations             = 1
Number of objective gradient evaluations             = 1
Number of equality constraint evaluations            = 0
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 0
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 0
Total seconds in IPOPT                               = 0.001

EXIT: Optimal Solution Found.
 
Ipopt 3.14.4: Optimal Solution Found
Process(`/Users/oscar/.julia/artifacts/3cfe57276b47c104286bd59245953723fbc1a938/bin/ipopt test.nl`, ProcessExited(0))

(ip) pkg> st
      Status `/private/tmp/ip/Project.toml`
  [9cc047cb] Ipopt_jll v300.1400.400+0

julia> exit()
(base) oscar@Oscars-MBP /tmp % julia +1.8 --project=ip
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.3 (2022-11-14)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

(ip) pkg> up
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
    Updating `/private/tmp/ip/Project.toml`
  [9cc047cb] ↑ Ipopt_jll v300.1400.400+0 ⇒ v300.1400.1000+0
    Updating `/private/tmp/ip/Manifest.toml`
  [9cc047cb] ↑ Ipopt_jll v300.1400.400+0 ⇒ v300.1400.1000+0
  [d7ed1dd3] ↑ MUMPS_seq_jll v5.4.1+0 ⇒ v500.500.101+0
⌅ [656ef2d0] ↑ OpenBLAS32_jll v0.3.12+1 ⇒ v0.3.17+0
  [0dad84c5] ~ ArgTools ⇒ v1.1.1
  [f43a241f] ~ Downloads ⇒ v1.6.0
  [7b1f6079] + FileWatching
  [b27032c2] ~ LibCURL ⇒ v0.6.3
  [ca575930] ~ NetworkOptions ⇒ v1.2.0
  [44cfe95a] ~ Pkg ⇒ v1.8.0
  [ea8e919c] ~ SHA ⇒ v0.7.0
  [fa267f1f] ~ TOML ⇒ v1.0.0
  [a4e569a6] ~ Tar ⇒ v1.10.1
  [e66e0078] ~ CompilerSupportLibraries_jll ⇒ v0.5.2+0
  [deac9b47] ~ LibCURL_jll ⇒ v7.84.0+0
  [29816b5a] ~ LibSSH2_jll ⇒ v1.10.2+0
  [c8ffd9c3] ~ MbedTLS_jll ⇒ v2.28.0+0
  [14a3606d] ~ MozillaCACerts_jll ⇒ v2022.2.1
  [4536629a] + OpenBLAS_jll v0.3.20+0
  [83775a58] ~ Zlib_jll ⇒ v1.2.12+3
  [8e850b90] + libblastrampoline_jll v5.1.1+0
  [8e850ede] ~ nghttp2_jll ⇒ v1.48.0+0
  [3f19e933] ~ p7zip_jll ⇒ v17.4.0+0
        Info Packages marked with ⌅ have new versions available but compatibility constraints restrict them from upgrading. To see why use `status --outdated -m`

(ip) pkg> st
Status `/private/tmp/ip/Project.toml`
  [9cc047cb] Ipopt_jll v300.1400.1000+0

julia> import Ipopt_jll

julia> Ipopt_jll.amplexe() do exe
           run(`$exe test.nl`)
       end


******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.10, running with linear solver MUMPS 5.5.1.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        1

ERROR: failed process: Process(`/Users/oscar/.julia/artifacts/6a0533fbc1b71b4c136e88c8407282c6eda28446/bin/ipopt test.nl`, ProcessSignaled(11)) [0]

Stacktrace:
  [1] pipeline_error
    @ ./process.jl:565 [inlined]
  [2] run(::Cmd; wait::Bool)
    @ Base ./process.jl:480
  [3] run
    @ ./process.jl:477 [inlined]
  [4] (::var"#1#2")(exe::String)
    @ Main ./REPL[6]:2
  [5] (::JLLWrappers.var"#2#3"{var"#1#2", String})()
    @ JLLWrappers ~/.julia/packages/JLLWrappers/QpMQW/src/runtime.jl:49
  [6] withenv(::JLLWrappers.var"#2#3"{var"#1#2", String}, ::Pair{String, String}, ::Vararg{Pair{String, String}})
    @ Base ./env.jl:172
  [7] withenv_executable_wrapper(f::Function, executable_path::String, PATH::String, LIBPATH::String, adjust_PATH::Bool, adjust_LIBPATH::Bool)
    @ JLLWrappers ~/.julia/packages/JLLWrappers/QpMQW/src/runtime.jl:48
  [8] invokelatest(::Any, ::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Base ./essentials.jl:729
  [9] invokelatest(::Any, ::Any, ::Vararg{Any})
    @ Base ./essentials.jl:726
 [10] amplexe(f::Function; adjust_PATH::Bool, adjust_LIBPATH::Bool)
    @ Ipopt_jll ~/.julia/packages/JLLWrappers/QpMQW/src/products/executable_generators.jl:21
 [11] amplexe(f::Function)
    @ Ipopt_jll ~/.julia/packages/JLLWrappers/QpMQW/src/products/executable_generators.jl:19
 [12] top-level scope
    @ REPL[6]:1
```

**Comment by amontoison** [comment]
@odow The issue is that `ipopt_jll` is linked with `LBT_jll` and we don't have any BLAS/LAPACK backend. In general we use Ipopt through Ipopt.jl and we load a `lp64` BLAS/LAPACK [here](https://github.com/jump-dev/Ipopt.jl/blob/master/src/Ipopt.jl#L15-L28) if we don't already have one.

**Comment by odow** [comment]
Ahhhhhh.

So we need

```Julia
    if VERSION >= v"1.8"
        config = LinearAlgebra.BLAS.lbt_get_config()
        if !any(lib -> lib.interface == :lp64, config.loaded_libs)
            LinearAlgebra.BLAS.lbt_forward(
                OpenBLAS32_jll.libopenblas_path;
                verbose = false,
                clear = false,
            )
        end
    end
```

**Comment by amontoison** [comment]
> Ahhhhhh.
> 
> So we need
> 
> ```julia
>     if VERSION >= v"1.8"
>         config = LinearAlgebra.BLAS.lbt_get_config()
>         if !any(lib -> lib.interface == :lp64, config.loaded_libs)
>             LinearAlgebra.BLAS.lbt_forward(
>                 OpenBLAS32_jll.libopenblas_path;
>                 verbose = false,
>                 clear = false,
>             )
>         end
>     end
> ```

Yes but we don't use the shared library here but the executable. I'm scared that the executable is standalone.

**Comment by odow** [comment]
Hmm. Yes that doesn't seem to have fixed it:
```Julia
julia> config = LinearAlgebra.BLAS.lbt_get_config()
LinearAlgebra.BLAS.LBTConfig
Libraries: 
└ [ILP64] libopenblas64_.0.3.20.dylib

julia> if !any(lib -> lib.interface == :lp64, config.loaded_libs)
                   LinearAlgebra.BLAS.lbt_forward(
                       OpenBLAS32_jll.libopenblas_path;
                       verbose = false,
                       clear = false,
                   )
               end
4860

julia> config = LinearAlgebra.BLAS.lbt_get_config()
LinearAlgebra.BLAS.LBTConfig
Libraries: 
├ [ILP64] libopenblas64_.0.3.20.dylib
└ [ LP64] libopenblas.0.3.17.dylib

julia> import Ipopt_jll

julia> Ipopt_jll.amplexe() do exe
           run(`$exe test.nl`)
       end


******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.10, running with linear solver MUMPS 5.5.1.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        1

ERROR: failed process: Process(`/Users/oscar/.julia/artifacts/6a0533fbc1b71b4c136e88c8407282c6eda28446/bin/ipopt test.nl`, ProcessSignaled(11)) [0]

```

**Comment by amontoison** [comment]
If the executable is standalone, we just need to add `OpenBLAS32_jll` as a dependency of `Ipopt_jll` and replace `-lblastrampoline` by `-lopenblas` but just for the executable.

**Comment by odow** [comment]
> If the executable is standalone, we need to add OpenBLAS32_jll as a dependency of Ipopt_jll and replace -lblastrampoline by -lopenblas but just for the executable

Yuck. I get that it's pretty hard to make work otherwise though.

cc @giordano

File: test/MOI_wrapper.jl [hunk 1]
```diff
@@ -138,15 +138,11 @@ function test_io(path)
     x = MOI.add_variable(model)
```
Comment (odow) [inline test/MOI_wrapper.jl:138]:
```suggestion
    MOI.set(model, MOI.RawOptimizerAttribute("print_level"), 1)
    x = MOI.add_variable(model)
```

**Comment by odow** [comment]
This adds `Ipopt_jll` as a full dependency so that people using AmplNLWriter and Ipopt_jll won't be broken. We can move it back to a test dependency once we fix Ipopt_jll.

**Pull request merged**